U4-11275 routPath in custom tree always needs a id

mvanhelmont opened the pull request:
Fixes http://issues.umbraco.org/issue/U4-11275

This is a simple fix for U4-11275 so that packages also can set routePath to something like a overview page. Like the new user section when you don't want to display children. Is a copy of /:section/:tree/:method/:id

dawoe commented:
Hi @mvanhelmont 

Thanks for this PR. We will try to review this in max 14 days. Maybe sooner.

One question. How are you setting the custo;m routePath in your tree ?

I just checked a  project of ours and for us the views are loaded correctly using this code in the treecontroller

```
            var treeNodeCollection = new TreeNodeCollection();

            var routePrefix = FormDataCollectionExtensions.GetValue<string>(queryStrings, "application")
                              + StringExtensions.EnsureStartsWith(this.TreeAlias, '/');

            if (id == Constants.System.Root.ToString())
            {
                // added google shopping node
                treeNodeCollection.Add(
                    this.CreateTreeNode(
                        "alias",
                        "-1",
                        queryStrings,
                        "Orders",
                        "icon-shopping-basket",
                        false,
                        routePrefix + "/orders/all"));  
                
            return treeNodeCollection;
        }
```


Dave (CPR Team)

mvanhelmont commented:
Hi @dawoe 

I do nodes.Add(this.CreateTreeNode("overview", id, queryStrings, "Products", "icon-folde", false, "customsection/customtree/overview"));

this then will try to load from the umbraco views folder /views/customtree/overview.html

if i do nodes.Add(this.CreateTreeNode("overview", id, queryStrings, "Products", "icon-folde", false, "customsection/customtree/overview/all"));

then to he loads it from /App_Plugins/customsection/backoffice/customtree/overview.html and passed a param id with all to the angular controller.

Hope it's clear now if not i will make some screenshots.

Thanks,
Marcel

dawoe commented:
I think you need to use this code to get the correct route prefix

 var routePrefix = FormDataCollectionExtensions.GetValue<string>(queryStrings, "application")
                              + StringExtensions.EnsureStartsWith(this.TreeAlias, '/');

mvanhelmont commented:
Hi @dawoe 

I've tried to use you code, but missing something so i needed to change it to:

```
var routePrefix = FormDataCollectionExtensions.GetValue<string>(queryStrings, "application") + StringExtensions.EnsureStartsWith(TreeAlias, '/');

nodes.Add(this.CreateTreeNode("overview", id, queryStrings, "Products", "icon-folde", false, routePrefix + "overview"));
```

Problem still remains because of not sending a extra id param. I still get the following message.

![error](https://user-images.githubusercontent.com/5310629/41117119-6b588d1e-6a8c-11e8-95e6-539ed6928faa.png)

```
var routePrefix = FormDataCollectionExtensions.GetValue<string>(queryStrings, "application") + StringExtensions.EnsureStartsWith(TreeAlias, '/');

nodes.Add(this.CreateTreeNode("overview", id, queryStrings, "Products", "icon-folde", false, routePrefix + "overview/somethingidontwant"));
```

this loads correct the overview.html view from the App_Plugins folder, but with the pull request the first line without somethingidontwant also will work for packages.

dawoe commented:
Hi @mvanhelmont 

I get what you want. The routing is now set up that an id param is required, even if you don't need for your view to work.

Problem the code I supplied is probably used in a lot of projects out there. And add first glance I think your code will break that.

I think this needs to be tested very well.

Dave

mvanhelmont commented:
Hi @dawoe 

I have changed your routePrefix code not in source (can't find the correct extension method only this one)

```
/// <summary>
        /// Returns the object based in the collection based on it's key. This does this with a conversion so if it doesn't convert a null object is returned.
        /// </summary>
        /// <typeparam name="T"></typeparam>
        /// <param name="items"></param>
        /// <param name="key"></param>
        /// <returns></returns>
        public static T GetValue<T>(this FormDataCollection items, string key)
        {
            var val = items.Get(key);
            if (string.IsNullOrEmpty(val)) return default(T);

            var converted = val.TryConvertTo<T>();
            return converted.Success 
                ? converted.Result 
                : default(T);
        }
```

But that's not the problem the only thing the routePrefix code does is get the correct section and tree alias en make the routePath like this customer section/treealias/overview

problem is indeed that id param is required in angular routes. So only changed angular routing a bit so it respect packages.

dawoe commented:
Hi @mvanhelmont 

I tested this, and everything seems to keep working. All core trees work and installed a package that installs  a custom tree. That works as well.

But i'm still in doubt if this is the best way to fix this.

@nul800sebastiaan maybe someone from HQ shed some light on this ?

nul800sebastiaan commented:
Thanks @mvanhelmont - I think it makes sense to do this, I don't really see a better way at the moment and from testing it out by a few people this seems to work nicely. 👍
